Implement communication with mfrc522 and reading UIDs of ISO/IEC 14443 compliant cards

adam-dej opened the pull request:
This version does not implement anticollision loop yet, therefore detects only one card in the field.

adam-dej committed:
Add MFRC522 register adresses and command constants

adam-dej committed:
Implement reading from and writing to mfrc522 registers

adam-dej committed:
Implement reading from and writing to mfrc522 registers

adam-dej committed:
Implement initialization and data transmission

adam-dej committed:
Add CRC calculation functions

adam-dej committed:
[WIP] Implement detecting iso14443a compliant cards

adam-dej committed:
Merge branch 'feature-mfrc522' of github.com:fmfi-svt-gate/mfrc522.py into feature-mfrc522

Conflicts:
	mfrc522/mfrc522.py

adam-dej committed:
Make card data transfer more efficient

adam-dej committed:
Implement reseting PICCs after the first REQA

adam-dej committed:
Minor code cleanup

adam-dej committed:
Add setup.py

adam-dej committed:
Fix typo in __calculate_crt function

adam-dej commented on mfrc522/mfrc522.py:
```diff
@@ -0,0 +1,206 @@
+        PCD_RESETPHASE      = 0x0F
+        PCD_CALCCRC         = 0x03
+
+    class Registers(Enum):
```
Those are constants, yet names are in CamelCase so that they are consistent with the mfrc522 datasheet.

AnotherKamila commented on mfrc522/iso14443com.py:
```diff
@@ -0,0 +1,55 @@
+import crcmod
+from .mfrc522 import *
```
Avoid `import *` -- explicitly name what is needed.

AnotherKamila commented on mfrc522/iso14443com.py:
```diff
@@ -0,0 +1,55 @@
+    if cascade_level != SEL_CASCADE_1 and \
+       cascade_level != SEL_CASCADE_2 and \
+       cascade_level != SEL_CASCADE_3:
+        return None
```
You indent like a farmer.

AnotherKamila commented on mfrc522/iso14443com.py:
```diff
@@ -0,0 +1,55 @@
+    return struct.pack('<H', __crc_func(data))
+
+def __perform_cascade(module, cascade_level):
+    if cascade_level != SEL_CASCADE_1 and \
```
Preferably rewrite this `if` to:

``` python
if cascade_level not in [ SEL_CASCADE_1, SEL_CASCADE_2, SEL_CASCADE_3 ]
```

...or even better make the enum and say `if cascade_level not in <enum>`.

AnotherKamila commented on mfrc522/iso14443com.py:
```diff
@@ -0,0 +1,55 @@
+from .mfrc522 import *
+import struct
+
+SEL_CASCADE_1   = 0x93
```
IMHO an enum would be nicer. ("Namespaces are one honking great idea -- let's do more of those!")

EDIT: Or not even an enum -- see below.

AnotherKamila commented on mfrc522/iso14443com.py:
```diff
@@ -0,0 +1,55 @@
+def __calculate_crc(data):
+    return struct.pack('<H', __crc_func(data))
+
+def __perform_cascade(module, cascade_level):
```
Rename `module` to `rfid_module` or something, the word "module" can mean just about anything.

AnotherKamila replied:
The recursive version would make sense if anticollision were implemented. But for now, what about:

``` python
CASCADE_LEVELS = [ 0x93, 0x95, 0x97 ]

def __perform_cascade(rfid_module):
    result = b''
    for cascade_level in CASCADE_LEVELS:
        # transmit ANTICOLLISION command
        uid_cln = module.transcieve(bytes([cascade_level, 0x20]))  # BTW, that's spelled "transceive".
        if len(uid_cln) != 4: raise ...  # Don't ignore errors!

        # transmit SELECT command
        data = bytes((cascade_level, 0x70)) + bytes(uid_cln)
        data += __calculate_crc(data)
        response = rfid_module.transcieve(data)

        result += uid_cln[1:4]

        if response[0] & 0x24 in [0x0, 0x20]:
            return result
        elif response[0] & 0x04:
            pass  # continue the cascade
        else:
            # Something's wrong if we get here.
            raise ...  # Say so.
```

AnotherKamila commented on mfrc522/iso14443com.py:
```diff
@@ -0,0 +1,55 @@
+    elif response[0] & 0x24 == 0:
+        return uid_cln[:4]
+
+def get_id(module):
```
Rename `module`.

AnotherKamila commented on mfrc522/iso14443com.py:
```diff
@@ -0,0 +1,55 @@
+    except NoTagError:
+        return None
+
+def are_cards_in_field(module):
```
Rename `module`.

AnotherKamila commented on mfrc522/iso14443com.py:
```diff
@@ -0,0 +1,55 @@
+        module.transcieve([0x26])  # REQA
+        module.write_register(MFRC522.Registers.BitFramingReg, 0x00)
+        return __perform_cascade(module, SEL_CASCADE_1)
+    except NoTagError:
```
Following our discussion, perhaps this shouldn't be here in the next version.

AnotherKamila commented on mfrc522/mfrc522.py:
```diff
@@ -0,0 +1,206 @@
+
+    def write_register(self, register, val):
+        if type(val) is int:
+            self.spi.transfer(bytes((register.value << 1, val)))
```
Why `<< 1`? Maybe I should RTFM, or maybe you should comment it -- I'll leave the decision up to you, but I'd prefer a comment :D

AnotherKamila commented on mfrc522/mfrc522.py:
```diff
@@ -0,0 +1,206 @@
+    def write_register(self, register, val):
+        if type(val) is int:
+            self.spi.transfer(bytes((register.value << 1, val)))
+        else:
```
Ugly handling of the "overloading" (two completely different + duplicating code paths when it's doing the same thing). What I'd prefer:

``` python
if type(val) is int:
    val = bytes([val])
<paste from L120 onwards -- now it is bytes>
```

AnotherKamila commented on mfrc522/mfrc522.py:
```diff
@@ -0,0 +1,206 @@
+    def read_register(self, register):
+        return self.spi.transfer(bytes(((register.value << 1) | 0x80, 0)))[1]
+
+    def read_register_data(self, register, amount):
```
When removing/merging these, avoid code duplication.

AnotherKamila commented on mfrc522/mfrc522.py:
```diff
@@ -0,0 +1,206 @@
+        return bytes((self.read_register(MFRC522.Registers.CRCResultRegH),
+                      self.read_register(MFRC522.Registers.CRCResultRegL)))
+
+    def transcieve(self, data):
```
The spelling is `transceive`.

AnotherKamila commented on setup.py:
```diff
@@ -0,0 +1,11 @@
+
+setup(name='libmfrc522',
+      description='A library for communicating with the MFRC522 RFID module',
+      version='1.0',
```
Really? :D

AnotherKamila commented on mfrc522/mfrc522.py:
```diff
@@ -0,0 +1,206 @@
+        response_bytes = self.read_register(MFRC522.Registers.FIFOLevelReg)
+
+        response = self.read_register_data(MFRC522.Registers.FIFODataReg,
+                                            response_bytes)
```
You indent like a farmer!

AnotherKamila commented:
Plus:

- magic numbers! That's actually not a good idea.
- Python has docstrings. Your code should have them, too.

TomiBelan commented on mfrc522/iso14443com.py:
```diff
@@ -0,0 +1,55 @@
+SEL_CASCADE_2   = 0x95
+SEL_CASCADE_3   = 0x97
+
+__crc_func = crcmod.mkCrcFun(0x11021, initCrc=0x6363)
```
Nitpick: radsej zacni s `_` miesto `__`. `__foo` sa pouziva iba na private metody v triedach (python spravi name mangling). Inde je iba `foo`, `_foo` a `__foo__`.
(Uh, tvarime sa medzinarodne? Nechce sa mi prekladat, pardon...)

TomiBelan commented on mfrc522/mfrc522.py:
```diff
@@ -0,0 +1,206 @@
+
+    MAX_LEN = 16
+
+    class Commands(Enum):
```
Commands a Registers asi mozes presunut von z MRFC522 - koniec koncov, plne meno teraz je `mrfc522.mrfc522.MRFC522.Registers`... a ak ich importuje aj niekto iny, `from .mrfc522 import Registers` je asi dostatocne konkretne.

TomiBelan commented on mfrc522/mfrc522.py:
```diff
@@ -0,0 +1,206 @@
+        request = bytes(((register.value << 1) | 0x80,)*(amount+1))
+        return self.spi.transfer(request)[1:]
+
+    def set_mask_in_register(self, reg, mask):
```
Len drobnost, ale zda sa ze niektore metody maju `register` a niektore `reg`?

TomiBelan commented on mfrc522/mfrc522.py:
```diff
@@ -0,0 +1,206 @@
+          * reset():          Pushes NRST signal of the reader HIGH,
+                              thus resetting it (and exiting the hard_powerdown)
+                              If the NRST is already high, this function shall
+                              pull it LOW and then HIGH again.
```
Nepoznam terminologiu, ale zda sa ze tu je (zamlcany prisudok) "pull it HIGH" aj ked vyssie je "push it HIGH".
Mozne fixy: a) "and then push it HIGH again." b) vsade "pull"? c) vsade "push"? d) vsade "set", alebo nejake ine slovo? e) neviem, mozno nic netreba menit?

adam-dej committed:
Refactor: Naming and structure cleanup

Rename functions and parameters, move Commands and Registers
outside of the MFRC522 class

adam-dej committed:
Change version to a more descriptive number

adam-dej committed:
Refactor: rewrite _perform_cascade function

adam-dej committed:
Refactor: rewrite functions for reading and writing registers

adam-dej committed:
Refactor: Explain magic constants and fix docstrings

AnotherKamila commented on mfrc522/iso14443com.py:
```diff
@@ -0,0 +1,72 @@
+
+class CardCommands(Enum):
+    REQA         = 0x26
+    ANTICOLL_n   = [0x93, 0x95, 0x97]
```
I would be happier if this were immutable.

AnotherKamila commented on mfrc522/iso14443com.py:
```diff
@@ -0,0 +1,72 @@
+        elif response[0] & 0x04:
+            # UID incomplete; continue the cascade
+            result += uid_cln[1:4]
+            pass
```
(add comment "continue", optionally without the `pass`)

adam-dej committed:
Refactor: review-induced fixes
